kernel/alarm.c

```diff
@@ -0,0 +1,251 @@
+ * deadline at or before the provided when.  Return a null pointer if
+ * the k_alarm is scheduled strictly after the provided when.
+ */
+static ALWAYS_INLINE struct k_alarm *alarm_due_by(sys_dnode_t *node,
+						  uint32_t when)
+{
+	struct k_alarm *cap = CONTAINER_OF(node, struct k_alarm, node);
+
+	if ((s32_t)(cap->deadline - when) > 0) {
+		return 0;
+	}
+	return cap;
+}
+
+/** Add alarm to the provided list before the first alarm that has a
+ * later deadline, or at the end if necessary.
+ *
+ * Return @true if the head of list changed as a result of this
+ * addition.
+ */
+static bool link_alarm(sys_dlist_t *list,
+		       struct k_alarm *alarm,
+		       u32_t now)
+{
+	sys_dnode_t *node = &alarm->node;
+	sys_dnode_t *cp = sys_dlist_peek_head(list);
+
+	while (cp) {
+		if (!alarm_due_by(cp, alarm->deadline)) {
+			break;
+		}
+		cp = sys_dlist_peek_next_no_check(list, cp);
+	}
+	if (cp) {
+		sys_dlist_insert(cp, node);
+	} else {
+		sys_dlist_append(list, node);
+	}
+	return sys_dlist_is_head(list, node);
+}
+
+void k_alarm_init(struct k_alarm *alarm,
+		  k_alarm_handler_t handler,
+		  k_alarm_cancel_t cancel,
+		  void *user_data)
+{
+	*alarm = Z_ALARM_INITIALIZER(handler, cancel, user_data);
+}
+
+int k_alarm_next_deadline_(u32_t *deadline)
+{
+	int rv = -1;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (!sys_dlist_is_empty(&ready_list)) {
+		rv = 0;
+		goto unlock;
+	}
+
+	sys_dnode_t *hp = sys_dlist_peek_head(&scheduled_list);
+
+	if (hp) {
+		struct k_alarm *ap = CONTAINER_OF(hp, struct k_alarm, node);
+		*deadline = ap->deadline;
+		rv = 1;
+	}
+unlock:
+	k_spin_unlock(&lock, key);
+	return rv;
+}
+
+int k_alarm_split_(u32_t now)
+{
+	int rv = 0;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	sys_dnode_t *cp = sys_dlist_peek_head(&scheduled_list);
+
+	while (cp) {
+		struct k_alarm *cap = alarm_due_by(cp, now);
+
+		if (!cap) {
+			break;
+		}
+		cap->state = K_ALARM_READY;
+		++rv;
+		cp = sys_dlist_peek_next_no_check(&scheduled_list, cp);
+	}
+	if ((rv == 0)
+	    && !sys_dlist_is_empty(&ready_list)) {
+		rv = -1;
+	}
+
+	if (cp) {
+		sys_dlist_split(&ready_list, &scheduled_list, cp);
+	} else {
+		sys_dlist_join(&ready_list, &scheduled_list);
+	}
+
+	k_spin_unlock(&lock, key);
+	return rv;
+}
+
+void k_alarm_process_ready_(void)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	do {
+		sys_dnode_t *cp = sys_dlist_get(&ready_list);
+
+		if (!cp) {
+			break;
+		}
+
+		struct k_alarm *cap = CONTAINER_OF(cp, struct k_alarm, node);
+
+		cap->state = K_ALARM_ACTIVE;
+		if (cap->handler_fn) {
+			k_spin_unlock(&lock, key);
+			cap->handler_fn(cap, cap->user_data);
+			key = k_spin_lock(&lock);
+		}
+		if (cap->state == K_ALARM_ACTIVE) {
+			cap->state = K_ALARM_UNSCHEDULED;
+		}
+	} while (true);
+	k_spin_unlock(&lock, key);
+}
+
+int z_impl_k_alarm_schedule(struct k_alarm *alarm,
+			    u32_t deadline,
+			    u32_t flags)
+{
+	int rv = -EINVAL;
+	bool update_deadline = false;
+
+	if (!alarm) {
```

Comment:
No, but user code gets here and if it isn't checked we'd get a hard fault.  I'm not a fan of assert since it fails at runtime and Zephyr doesn't have a robust infrastructure for handling failure safely.  This way nothing gets corrupted and the application has a chance to detect what went wrong.